// exupery-core-alg/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.data_processing = exports.block = exports.dictionary_literal = exports.list_literal = exports.deprecated_panic = exports.ss = exports.au = exports.cc = exports.not_set = exports.set = void 0;
/**
 * these functions coming from core-internals should be exposed for library development
 */
var exupery_core_internals_1 = require("exupery-core-internals");
Object.defineProperty(exports, "set", { enumerable: true, get: function () { return exupery_core_internals_1.set; } });
Object.defineProperty(exports, "not_set", { enumerable: true, get: function () { return exupery_core_internals_1.not_set; } });
Object.defineProperty(exports, "cc", { enumerable: true, get: function () { return exupery_core_internals_1.cc; } });
Object.defineProperty(exports, "au", { enumerable: true, get: function () { return exupery_core_internals_1.au; } });
Object.defineProperty(exports, "ss", { enumerable: true, get: function () { return exupery_core_internals_1.ss; } });
/**
 * panic should no longer be used,
 * if it is used in a transformation, this means it is not a transformation but a refinement
 */
Object.defineProperty(exports, "deprecated_panic", { enumerable: true, get: function () { return exupery_core_internals_1.panic; } });
Object.defineProperty(exports, "list_literal", { enumerable: true, get: function () { return exupery_core_internals_1.list_literal; } });
Object.defineProperty(exports, "dictionary_literal", { enumerable: true, get: function () { return exupery_core_internals_1.dictionary_literal; } });
Object.defineProperty(exports, "block", { enumerable: true, get: function () { return exupery_core_internals_1.block; } });
Object.defineProperty(exports, "data_processing", { enumerable: true, get: function () { return exupery_core_internals_1.data_processing; } });
__exportStar(require("./build"), exports);
__exportStar(require("./group"), exports);
__exportStar(require("./create_refinement_context"), exports);
__exportStar(require("./text_to_character_list"), exports);
__exportStar(require("./integer"), exports);
__exportStar(require("./temp_julian"), exports);
