// pareto-fountain-pen/dist/implementation/generated/pareto/schemas/block/serialize.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Directory = exports.Node = exports.Group = exports.Group_Part = exports.Block = exports.Block_Part = void 0;
const _i_serialize = __importStar(require("../../generic/serialize"));
const _i_marshall = __importStar(require("./marshall"));
const Block_Part = ($, $p) => _i_serialize.Document(_i_marshall.Block_Part($, {
    'value serializers': $p['value serializers'],
}));
exports.Block_Part = Block_Part;
const Block = ($, $p) => _i_serialize.Document(_i_marshall.Block($, {
    'value serializers': $p['value serializers'],
}));
exports.Block = Block;
const Group_Part = ($, $p) => _i_serialize.Document(_i_marshall.Group_Part($, {
    'value serializers': $p['value serializers'],
}));
exports.Group_Part = Group_Part;
const Group = ($, $p) => _i_serialize.Document(_i_marshall.Group($, {
    'value serializers': $p['value serializers'],
}));
exports.Group = Group;
const Node = ($, $p) => _i_serialize.Document(_i_marshall.Node($, {
    'value serializers': $p['value serializers'],
}));
exports.Node = Node;
const Directory = ($, $p) => _i_serialize.Document(_i_marshall.Directory($, {
    'value serializers': $p['value serializers'],
}));
exports.Directory = Directory;
