// exupery-resources/dist/implementation/generated/pareto/generic/parse/astn_parse_generic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.create_astn_token_iterator = exports.throw_unexpected_token = exports.throw_lexer_error = exports.Parse_Error_Class = void 0;
class Parse_Error_Class {
    type;
    range;
    constructor(type, range) {
        this.type = type;
        this.range = range;
    }
}
exports.Parse_Error_Class = Parse_Error_Class;
const throw_parse_error = (type, range) => {
    throw new Parse_Error_Class(type, range);
};
const throw_lexer_error = (type, range) => {
    throw new Parse_Error_Class(['lexer', type], range);
};
exports.throw_lexer_error = throw_lexer_error;
const throw_unexpected_token = (found, expected) => {
    return throw_parse_error(['parser', {
            'expected': expected,
            'cause': ['unexpected token', {
                    'found': found.type,
                }]
        }], {
        'start': found.start,
        'end': found.end
    });
};
exports.throw_unexpected_token = throw_unexpected_token;
const create_astn_token_iterator = ($, end) => {
    let position = 0;
    return {
        'get required token': (pet) => {
            return $.__get_element_at(position).transform(($) => $, () => throw_parse_error(['parser', {
                    'expected': pet,
                    'cause': ['missing token', null]
                }], {
                'start': end,
                'end': end,
            }));
        },
        'consume token': () => {
            position += 1;
        },
    };
};
exports.create_astn_token_iterator = create_astn_token_iterator;
