// exupery-core-data/dist/shorthands/unresolved_data.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrap_stack_reference = exports.wrap_reference = exports.wrap_state_group = exports.wrap_list = exports.wrap_dictionary = exports.to_raw_array = void 0;
const _ei = __importStar(require("exupery-core-internals"));
const get_location_info_1 = require("../get_location_info");
const depth = 1;
const to_raw_array = ($) => $.__get_raw_copy();
exports.to_raw_array = to_raw_array;
const wrap_dictionary = ($) => {
    const location = (0, get_location_info_1.get_location_info)(depth + 1);
    function is_normal($) {
        return $.__get_number_of_entries !== undefined && typeof $.__get_number_of_entries === "function";
    }
    if (is_normal($)) {
        return {
            'location': location,
            'dictionary': $.map(($) => ({
                'location': location,
                'entry': $,
            }))
        };
    }
    else {
        return {
            'location': location,
            'dictionary': _ei.dictionary_literal($).map(($) => ({
                'location': location,
                'entry': $,
            }))
        };
    }
};
exports.wrap_dictionary = wrap_dictionary;
const wrap_list = ($) => {
    const location = (0, get_location_info_1.get_location_info)(depth + 1);
    const decorated = $ instanceof Array
        ? _ei.list_literal($)
        : $;
    if (!(decorated.__for_each instanceof Function)) {
        throw new Error("invalid input in 'wrap_list'");
    }
    return {
        'location': location,
        'list': decorated.map(($) => ({
            'location': location,
            'element': $,
        }))
    };
};
exports.wrap_list = wrap_list;
const wrap_state_group = ($) => {
    return {
        'location': (0, get_location_info_1.get_location_info)(depth + 1),
        'state group': $,
    };
};
exports.wrap_state_group = wrap_state_group;
const wrap_reference = ($) => {
    return {
        'location': (0, get_location_info_1.get_location_info)(depth + 1),
        'key': $,
    };
};
exports.wrap_reference = wrap_reference;
const wrap_stack_reference = (name) => {
    return {
        'location': (0, get_location_info_1.get_location_info)(depth + 1),
        'key': name,
    };
};
exports.wrap_stack_reference = wrap_stack_reference;
