// exupery-core-internals/dist/create_staging_result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.__create_staging_result = __create_staging_result;
class Staging_Result_Class {
    constructor(executer) {
        this.executer = executer;
    }
    transform(transformer) {
        return new Staging_Result_Class((on_result, on_error) => {
            this.executer(($) => {
                on_result(transformer($));
            }, on_error);
        });
    }
    transform_error_temp(error_transformer) {
        return new Staging_Result_Class((on_result, on_error) => {
            this.executer(on_result, ($) => {
                on_error(error_transformer($));
            });
        });
    }
    stage_without_error_transformation(stager) {
        return new Staging_Result_Class((on_result, on_error) => {
            this.executer(($) => {
                stager($).__extract_data(on_result, on_error);
            }, on_error);
        });
    }
    stage(stager, error_transformer) {
        return new Staging_Result_Class((on_result, on_error) => {
            this.executer(($) => {
                stager($).__extract_data(on_result, (stager_error) => {
                    on_error(error_transformer(stager_error));
                });
            }, on_error);
        });
    }
    rework_error_temp(error_reworker, rework_error_transformer) {
        return new Staging_Result_Class((on_result, on_error) => {
            this.executer(on_result, ($) => {
                error_reworker($).__extract_data((new_target_error) => {
                    on_error(new_target_error);
                }, (rework_error) => {
                    on_error(rework_error_transformer(rework_error));
                });
            });
        });
    }
    __extract_data(on_result, on_error) {
        this.executer(on_result, on_error);
    }
}
function __create_staging_result(executer) {
    return new Staging_Result_Class(executer);
}
