// pareto-fountain-pen/dist/implementation/generated/pareto/generic/parse/ast.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Value = exports.Key_Value_Pairs = exports.Elements = exports.Document = exports.String = exports.Structural_Token = void 0;
const _ea = __importStar(require("exupery-core-alg"));
const pg = __importStar(require("./astn_parse_generic"));
//this file contains the parser functionality, each function return a type from the 'ast' schema
const Structural_Token = (token) => {
    return {
        'trailing trivia': token['trailing trivia'],
        'range': {
            'start': token['start'],
            'end': token['end']
        }
    };
};
exports.Structural_Token = Structural_Token;
const String = (token_iterator) => {
    const token = token_iterator['get required token'](_ea.list_literal([['a string', null]]));
    if (token.type[0] !== 'string') {
        return pg.throw_unexpected_token(token, _ea.list_literal([['a string', null]]));
    }
    token_iterator['consume token']();
    return {
        'range': {
            'start': token['start'],
            'end': token['end']
        },
        'value': token.type[1].value,
        'type': token.type[1].type,
        'trailing trivia': token['trailing trivia'],
    };
};
exports.String = String;
const Document = (token_iterator) => {
    return {
        'header': _ea.block(() => {
            const token = token_iterator['get required token'](_ea.list_literal([['!', null], ['a value', null]]));
            if (token.type[0] !== '!') {
                return _ea.not_set();
            }
            token_iterator['consume token']();
            return _ea.set({
                '!': (0, exports.Structural_Token)(token),
                'value': (0, exports.Value)(token_iterator)
            });
        }),
        'content': (0, exports.Value)(token_iterator)
    };
};
exports.Document = Document;
const Elements = (token_iterator, end_reached, end_token) => {
    return _ea.build_list(($i) => {
        while (true) {
            const current_token = token_iterator['get required token'](_ea.list_literal([end_token, ['a value', null]]));
            if (end_reached(current_token.type)) {
                return;
            }
            $i['add element']({
                'value': (0, exports.Value)(token_iterator),
            });
        }
    });
};
exports.Elements = Elements;
const Key_Value_Pairs = (token_iterator, end_reached, end_token) => {
    return _ea.build_list(($i) => {
        while (true) {
            const current_token = token_iterator['get required token'](_ea.list_literal([end_token, ['a string', null]]));
            if (end_reached(current_token.type)) {
                return;
            }
            $i['add element']({
                'key': (0, exports.String)(token_iterator),
                'value': _ea.block(() => {
                    const candidate_colon = token_iterator['get required token'](_ea.list_literal([['a string', null], [':', null], end_token]));
                    if (candidate_colon.type[0] !== ':') {
                        return _ea.not_set();
                    }
                    token_iterator['consume token']();
                    return _ea.set({
                        ':': (0, exports.Structural_Token)(candidate_colon),
                        'value': (0, exports.Value)(token_iterator)
                    });
                }),
                ',': _ea.not_set()
            });
        }
    });
};
exports.Key_Value_Pairs = Key_Value_Pairs;
const Value = (token_iterator) => {
    const token = token_iterator['get required token'](_ea.list_literal([['a value', null]]));
    return _ea.cc(token.type, ($) => {
        switch ($[0]) {
            case 'string': return _ea.ss($, ($) => {
                return ['string', (0, exports.String)(token_iterator)];
            });
            case '{': return _ea.ss($, ($) => {
                token_iterator['consume token']();
                return ['indexed collection', ['dictionary', {
                            '{': (0, exports.Structural_Token)(token),
                            'entries': (0, exports.Key_Value_Pairs)(token_iterator, ($) => $[0] === '}', ['}', null]),
                            '}': _ea.block(() => {
                                const current_token = token_iterator['get required token'](_ea.list_literal([['}', null]]));
                                token_iterator['consume token']();
                                return (0, exports.Structural_Token)(current_token);
                            })
                        }]];
            });
            case '(': return _ea.ss($, ($) => {
                token_iterator['consume token']();
                return ['indexed collection', ['verbose group', {
                            '(': (0, exports.Structural_Token)(token),
                            'entries': (0, exports.Key_Value_Pairs)(token_iterator, ($) => $[0] === ')', [')', null]),
                            ')': _ea.block(() => {
                                const current_token = token_iterator['get required token'](_ea.list_literal([[')', null]]));
                                token_iterator['consume token']();
                                return (0, exports.Structural_Token)(current_token);
                            })
                        }]];
            });
            case '[': return _ea.ss($, ($) => {
                token_iterator['consume token']();
                return ['ordered collection', ['list', {
                            '[': (0, exports.Structural_Token)(token),
                            'elements': (0, exports.Elements)(token_iterator, ($) => $[0] === ']', [']', null]),
                            ']': _ea.block(() => {
                                const current_token = token_iterator['get required token'](_ea.list_literal([[']', null]]));
                                token_iterator['consume token']();
                                return (0, exports.Structural_Token)(current_token);
                            }),
                        }]];
            });
            case '<': return _ea.ss($, ($) => {
                token_iterator['consume token']();
                return ['ordered collection', ['concise group', {
                            '<': (0, exports.Structural_Token)(token),
                            'elements': (0, exports.Elements)(token_iterator, ($) => $[0] === '>', ['>', null]),
                            '>': _ea.block(() => {
                                const current_token = token_iterator['get required token'](_ea.list_literal([['>', null]]));
                                token_iterator['consume token']();
                                return (0, exports.Structural_Token)(current_token);
                            }),
                        }]];
            });
            case '|': return _ea.ss($, ($) => {
                token_iterator['consume token']();
                const token = token_iterator['get required token'](_ea.list_literal([['a value', null], ['#', null]]));
                return ['tagged value', {
                        '|': (0, exports.Structural_Token)(token),
                        'state': (0, exports.String)(token_iterator),
                        'value': (0, exports.Value)(token_iterator)
                    }];
            });
            case '*': return _ea.ss($, ($) => {
                token_iterator['consume token']();
                return ['set optional value', {
                        '*': (0, exports.Structural_Token)(token),
                        'value': (0, exports.Value)(token_iterator)
                    }];
            });
            case '~': return _ea.ss($, ($) => {
                token_iterator['consume token']();
                return ['not set', {
                        '~': (0, exports.Structural_Token)(token),
                    }];
            });
            //unexpected tokens
            case '#': return _ea.ss($, ($) => {
                token_iterator['consume token']();
                return pg.throw_unexpected_token(token, _ea.list_literal([
                    ['a value', null]
                ]));
            });
            case '@': return _ea.ss($, ($) => {
                token_iterator['consume token']();
                return pg.throw_unexpected_token(token, _ea.list_literal([
                    ['a value', null]
                ]));
            });
            case '!': return _ea.ss($, ($) => {
                token_iterator['consume token']();
                return pg.throw_unexpected_token(token, _ea.list_literal([
                    ['a value', null]
                ]));
            });
            case ':': return _ea.ss($, ($) => {
                token_iterator['consume token']();
                return pg.throw_unexpected_token(token, _ea.list_literal([
                    ['a value', null]
                ]));
            });
            case ')': return _ea.ss($, ($) => {
                token_iterator['consume token']();
                return pg.throw_unexpected_token(token, _ea.list_literal([
                    ['a value', null]
                ]));
            });
            case '>': return _ea.ss($, ($) => {
                token_iterator['consume token']();
                return pg.throw_unexpected_token(token, _ea.list_literal([
                    ['a value', null]
                ]));
            });
            case ']': return _ea.ss($, ($) => {
                token_iterator['consume token']();
                return pg.throw_unexpected_token(token, _ea.list_literal([
                    ['a value', null]
                ]));
            });
            case '}': return _ea.ss($, ($) => {
                token_iterator['consume token']();
                return pg.throw_unexpected_token(token, _ea.list_literal([
                    ['a value', null]
                ]));
            });
            default: return _ea.au($[0]);
        }
    });
};
exports.Value = Value;
