// pareto-fountain-pen/dist/implementation/generated/pareto/schemas/block/unmarshall.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Directory = exports.Node = exports.Group = exports.Group_Part = exports.Block = exports.Block_Part = void 0;
const _pa = __importStar(require("exupery-core-alg"));
const _i_generic = __importStar(require("../../generic/unmarshall"));
const Block_Part = ($, $p) => _i_generic.process_unconstrained_state_group($, {
    'states': _pa.dictionary_literal({
        'snippet': ($) => ['snippet', _i_generic.process_text($, null)],
        'indent': ($) => ['indent', (0, exports.Group)($, {
                'value deserializers': $p['value deserializers'],
            })],
        'sub block': ($) => ['sub block', (0, exports.Block)($, {
                'value deserializers': $p['value deserializers'],
            })],
        'optional': ($) => ['optional', _i_generic.process_optional($, {
                'value': ($) => (0, exports.Block_Part)($, {
                    'value deserializers': $p['value deserializers'],
                }),
            })],
        'nothing': ($) => ['nothing', _i_generic.process_nothing($, null)],
    }),
});
exports.Block_Part = Block_Part;
const Block = ($, $p) => _i_generic.process_unconstrained_list($, {
    'value': ($) => (0, exports.Block_Part)($, {
        'value deserializers': $p['value deserializers'],
    }),
});
exports.Block = Block;
const Group_Part = ($, $p) => _i_generic.process_unconstrained_state_group($, {
    'states': _pa.dictionary_literal({
        'nested block': ($) => ['nested block', (0, exports.Block)($, {
                'value deserializers': $p['value deserializers'],
            })],
        'block': ($) => ['block', _i_generic.process_text($, null)],
        'sub group': ($) => ['sub group', (0, exports.Group)($, {
                'value deserializers': $p['value deserializers'],
            })],
        'optional': ($) => ['optional', _i_generic.process_optional($, {
                'value': ($) => (0, exports.Group_Part)($, {
                    'value deserializers': $p['value deserializers'],
                }),
            })],
        'nothing': ($) => ['nothing', _i_generic.process_nothing($, null)],
    }),
});
exports.Group_Part = Group_Part;
const Group = ($, $p) => _i_generic.process_unconstrained_list($, {
    'value': ($) => (0, exports.Group_Part)($, {
        'value deserializers': $p['value deserializers'],
    }),
});
exports.Group = Group;
const Node = ($, $p) => _i_generic.process_unconstrained_state_group($, {
    'states': _pa.dictionary_literal({
        'file': ($) => ['file', (0, exports.Group)($, {
                'value deserializers': $p['value deserializers'],
            })],
        'directory': ($) => ['directory', (0, exports.Directory)($, {
                'value deserializers': $p['value deserializers'],
            })],
    }),
});
exports.Node = Node;
const Directory = ($, $p) => _i_generic.process_unconstrained_dictionary($, {
    'value': ($) => (0, exports.Node)($, {
        'value deserializers': $p['value deserializers'],
    }),
});
exports.Directory = Directory;
