// exupery-resources/dist/implementation/generated/pareto/generic/unmarshall.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.process_derived_reference = exports.process_stack_reference = exports.process_selected_reference = exports.process_nothing = exports.process_optional = exports.process_unconstrained_list = exports.process_unresolved_list = exports.process_text = exports.process_boolean = exports.process_number = exports.process_unconstrained_dictionary = exports.process_unresolved_dictionary = exports.get_entry = exports.process_group = exports.process_unresolved_state_group = exports.process_unconstrained_state_group = void 0;
const _ea = __importStar(require("exupery-core-alg"));
const process_unconstrained_state_group = ($, $p) => {
    return _ea.cc($, ($) => {
        switch ($[0]) {
            case 'tagged value': return _ea.ss($, ($) => {
                const data = $.value;
                return $p.states.__get_entry($.state.value).transform(($) => $(data), () => _ea.deprecated_panic(`Unexpected state: ${$.state.value}`));
            });
            default: return _ea.deprecated_panic(`Unexpected type for state group: ${$[0]}`);
        }
    });
};
exports.process_unconstrained_state_group = process_unconstrained_state_group;
const process_unresolved_state_group = ($, $p) => {
    return _ea.cc($, ($) => {
        switch ($[0]) {
            case 'tagged value': return _ea.ss($, ($) => {
                const data = $.value;
                return {
                    'location': $["|"].range,
                    'state group': $p.states.__get_entry($.state.value).transform(($) => $(data), () => _ea.deprecated_panic(`Unexpected state: ${$.state.value}`))
                };
            });
            default: return _ea.deprecated_panic(`Unexpected type for state group: ${$[0]}`);
        }
    });
};
exports.process_unresolved_state_group = process_unresolved_state_group;
const process_group = ($, $p) => {
    return _ea.cc($, ($) => {
        switch ($[0]) {
            case 'indexed collection': return _ea.ss($, ($) => _ea.cc($, ($) => {
                switch ($[0]) {
                    case 'verbose group': return _ea.ss($, ($) => {
                        return $p.properties(_ea.deprecated_build_dictionary(($i) => {
                            $.entries.__for_each(($) => {
                                $i['add entry']($.key.value, $.value.transform(($) => $.value, () => _ea.deprecated_panic(`no value for property: ${$.key.value}`)));
                            });
                        }));
                    });
                    default: return _ea.deprecated_panic(`Unexpected type for group: ${$[0]}`);
                }
            }));
            default: return _ea.deprecated_panic(`Unexpected type for group: ${$[0]}`);
        }
    });
};
exports.process_group = process_group;
const get_entry = ($, $p) => {
    return $.__get_entry($p.key).transform(($) => $, () => _ea.deprecated_panic(`no such entry: ${$p.key}`));
};
exports.get_entry = get_entry;
const process_unresolved_dictionary = ($, $p) => {
    return _ea.cc($, ($) => {
        switch ($[0]) {
            case 'indexed collection': return _ea.ss($, ($) => _ea.cc($, ($) => {
                switch ($[0]) {
                    case 'dictionary': return _ea.ss($, ($) => {
                        return {
                            'location': {
                                'start': $["{"].range.start,
                                'end': $["}"].range.end,
                            },
                            'dictionary': _ea.deprecated_build_dictionary(($i) => {
                                $.entries.__for_each(($) => {
                                    const key_location = $.key.range;
                                    $i['add entry']($.key.value, $.value.transform(($) => ({
                                        'location': key_location,
                                        'entry': $p.value($.value),
                                    }), () => _ea.deprecated_panic(`no value for property: ${$.key.value}`)));
                                });
                            })
                        };
                    });
                    default: return _ea.deprecated_panic(`Unexpected type for dictionary: ${$[0]}`);
                }
            }));
            default: return _ea.deprecated_panic(`Unexpected type for dictionary: ${$[0]}`);
        }
    });
};
exports.process_unresolved_dictionary = process_unresolved_dictionary;
const process_unconstrained_dictionary = ($, $p) => {
    return _ea.cc($, ($) => {
        switch ($[0]) {
            case 'indexed collection': return _ea.ss($, ($) => _ea.cc($, ($) => {
                switch ($[0]) {
                    case 'dictionary': return _ea.ss($, ($) => {
                        return _ea.deprecated_build_dictionary(($i) => {
                            $.entries.__for_each(($) => {
                                $i['add entry']($.key.value, $.value.transform(($) => $p.value($.value), () => _ea.deprecated_panic(`no value for property: ${$.key.value}`)));
                            });
                        });
                    });
                    default: return _ea.deprecated_panic(`Unexpected type for dictionary: ${$[0]}`);
                }
            }));
            default: return _ea.deprecated_panic(`Unexpected type for dictionary: ${$[0]}`);
        }
    });
};
exports.process_unconstrained_dictionary = process_unconstrained_dictionary;
const process_number = ($, $p) => {
    return _ea.cc($, ($) => {
        switch ($[0]) {
            case 'string': return _ea.ss($, ($) => $p.deserializer($.value, null));
            default: return _ea.deprecated_panic(`Unexpected type for number: ${$[0]}`);
        }
    });
};
exports.process_number = process_number;
const process_boolean = ($, $p) => {
    return _ea.cc($, ($) => {
        switch ($[0]) {
            case 'string': return _ea.ss($, ($) => $p.deserializer($.value, null));
            default: return _ea.deprecated_panic(`Unexpected type for boolean: ${$[0]}`);
        }
    });
};
exports.process_boolean = process_boolean;
const process_text = ($, $p) => {
    return _ea.cc($, ($) => {
        switch ($[0]) {
            case 'string': return _ea.ss($, ($) => $.value);
            default: return _ea.deprecated_panic(`Unexpected type for text: ${$[0]}`);
        }
    });
};
exports.process_text = process_text;
const process_unresolved_list = ($, $p) => {
    return _ea.cc($, ($) => {
        switch ($[0]) {
            case 'ordered collection': return _ea.ss($, ($) => _ea.cc($, ($) => {
                switch ($[0]) {
                    case 'list': return _ea.ss($, ($) => {
                        const temp = $["["];
                        return {
                            'location': $["["].range,
                            'list': $["elements"].map(($) => ({
                                'location': temp.range,
                                'element': $p.value($.value)
                            }))
                        };
                    });
                    default: return _ea.deprecated_panic(`Unexpected type for list: ${$[0]}`);
                }
            }));
            default: return _ea.deprecated_panic(`Unexpected type for list: ${$[0]}`);
        }
    });
};
exports.process_unresolved_list = process_unresolved_list;
const process_unconstrained_list = ($, $p) => {
    return _ea.cc($, ($) => {
        switch ($[0]) {
            case 'ordered collection': return _ea.ss($, ($) => _ea.cc($, ($) => {
                switch ($[0]) {
                    case 'list': return _ea.ss($, ($) => {
                        return $["elements"].map(($) => $p.value($.value));
                    });
                    default: return _ea.deprecated_panic(`Unexpected type for list: ${$[0]}`);
                }
            }));
            default: return _ea.deprecated_panic(`Unexpected type for list: ${$[0]}`);
        }
    });
};
exports.process_unconstrained_list = process_unconstrained_list;
const process_optional = ($, $p) => {
    return _ea.cc($, ($) => {
        switch ($[0]) {
            case 'not set': return _ea.ss($, ($) => _ea.not_set());
            case 'set optional value': return _ea.ss($, ($) => _ea.set($p.value($.value)));
            default: return _ea.deprecated_panic(`Unexpected type for nothing: ${$[0]}`);
        }
    });
};
exports.process_optional = process_optional;
const process_nothing = ($, $p) => {
    return _ea.cc($, ($) => {
        switch ($[0]) {
            case 'not set': return _ea.ss($, ($) => null);
            default: return _ea.deprecated_panic(`Unexpected type for nothing: ${$[0]}`);
        }
    });
};
exports.process_nothing = process_nothing;
const process_selected_reference = ($, $p) => {
    return _ea.cc($, ($) => {
        switch ($[0]) {
            case 'string': return _ea.ss($, ($) => ({
                'key': $.value,
                'location': $.range,
            }));
            default: return _ea.deprecated_panic(`Unexpected type for selected reference key: ${$[0]}`);
        }
    });
};
exports.process_selected_reference = process_selected_reference;
const process_stack_reference = ($, $p) => {
    return _ea.cc($, ($) => {
        switch ($[0]) {
            case 'string': return _ea.ss($, ($) => ({
                'key': $.value,
                'location': $.range,
            }));
            default: return _ea.deprecated_panic(`Unexpected type for selected reference key: ${$[0]}`);
        }
    });
};
exports.process_stack_reference = process_stack_reference;
const process_derived_reference = ($, $p) => {
    return _ea.cc($, ($) => {
        switch ($[0]) {
            case 'not set': return _ea.ss($, ($) => null);
            default: return _ea.deprecated_panic(`Unexpected type for derived reference: ${$[0]}`);
        }
    });
};
exports.process_derived_reference = process_derived_reference;
